pkg/k8s/apis/cilium.io/v2alpha1/bgp_cluster_types.go

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package v2alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	slimv1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:categories={cilium,ciliumbgp},singular="ciliumbgpclusterconfig",path="ciliumbgpclusterconfigs",scope="Cluster",shortName={bgpcluster}
+// +kubebuilder:printcolumn:JSONPath=".metadata.creationTimestamp",name="Age",type=date
+// +kubebuilder:storageversion
+
+// CiliumBGPClusterConfig is the Schema for the CiliumBGPClusterConfig API
+type CiliumBGPClusterConfig struct {
+	// +deepequal-gen=false
+	metav1.TypeMeta `json:",inline"`
+	// +deepequal-gen=false
+	metav1.ObjectMeta `json:"metadata"`
+
+	// Spec defines the desired cluster configuration of the BGP control plane.
+	Spec CiliumBGPClusterConfigSpec `json:"spec,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=false
+// +deepequal-gen=false
+
+// CiliumBGPClusterConfigList is a list of CiliumBGPClusterConfig objects.
+type CiliumBGPClusterConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	// Items is a list of CiliumBGPClusterConfig.
+	Items []CiliumBGPClusterConfig `json:"items"`
+}
+
+type CiliumBGPClusterConfigSpec struct {
+	// NodeSelector selects a group of nodes where this BGP Cluster
+	// config applies.
+	//
+	// If empty / nil this config applies to all nodes.
+	//
+	// +kubebuilder:validation:Optional
+	NodeSelector *slimv1.LabelSelector `json:"nodeSelector,omitempty"`
+
+	// A list of CiliumBGPInstance(s) which instructs
+	// the BGP control plane how to instantiate virtual BGP routers.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	// +listType=map
+	// +listMapKey=localASN
+	BGPInstances []CiliumBGPInstance `json:"bgpInstances"`
+}
+
+type CiliumBGPInstance struct {
+	// LocalASN is the ASN of this virtual router.
+	// Supports extended 32bit ASNs
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=4294967295
+	LocalASN int64 `json:"localASN"`
+
+	// Peers is a list of neighboring BGP peers for this virtual router
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	// +listType=map
+	// +listMapKey=peerAddress
+	Peers []CiliumBGPPeer `json:"peers"`
+}
+
+type CiliumBGPPeer struct {
+	// PeerAddress is the IP address of the neighbor.
+	// Supports IPv4 and IPv6 addresses.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Format=ip
```

Comment:
I see `ipv4` and `ipv6` as supported formats in the [go-openapi toolkit](https://pkg.go.dev/github.com/go-openapi/strfmt) but not `ip`. According to `kubernetes/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go`, unknown formats are ignored. Where is this format documented?

Reply:
Ah you're right, good catch! I wonder if there is way for conditional check. Or we have to go regex route to validate v4/v6 address.

Reply:
Unless we have separate IPv4 and IPv5 fields, we will need to use regex or validate within the controller.

Reply:
`ip` format will not be supported until it's added to [JSON schema validation](https://json-schema.org/draft/2020-12/json-schema-validation#name-ip-addresses).

Reply:
Added regex pattern match.